source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc

```diff
@@ -18,7 +18,69 @@ namespace Extensions {
 namespace HttpFilters {
 namespace HeaderToMetadataFilter {
 
-Rule::Rule(const std::string& header, const ProtoRule& rule) : header_(header), rule_(rule) {
+// extract the value of the header
+absl::optional<std::string> HeaderValueSelector::extract(Http::HeaderMap& map) const {
+  const Http::HeaderEntry* header_entry = map.get(header_);
+  if (header_entry == nullptr) {
+    return absl::nullopt; // no value
+  }
+  // catch the value in the header
+  absl::optional<std::string> value = std::string(header_entry->value().getStringView());
+  if (remove_) {
+    map.remove(header_); // remove the header
+  }
+  return value;
+}
+
+// extract the value of the key from the cookie header
+absl::optional<std::string> CookieValueSelector::extract(Http::HeaderMap& map) const {
+  std::string value = Envoy::Http::Utility::parseCookieValue(map, cookie_);
+  if (!value.empty()) {
+    return absl::optional<std::string>(value);
+  }
+  return absl::nullopt; // no value
```

Comment:
Hmm do we want to support removing the cookie (as we do for header)?

Reply:
If we don't -- we need to call it out since there's nothing in the proto definition that says remove doesn't work for cookies.

Reply:
nvm -- just saw there is a comment. we are good.

Reply:
nit: drop the comment too